src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,900 @@
+ *    limitations under the License.
+ */
+
+#include "time-synchronization-server.h"
+#include "TimeSyncManager.h"
+#include "time-synchronization-delegate.h"
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandler.h>
+#include <app/EventLogging.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SortUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#include <app-common/zap-generated/cluster-enums.h>
+
+#include <system/SystemClock.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::DeviceLayer;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TimeSynchronization;
+using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+using chip::TimeSyncDataProvider;
+using chip::Protocols::InteractionModel::Status;
+
+// -----------------------------------------------------------------------------
+// Delegate Implementation
+
+using chip::app::Clusters::TimeSynchronization::Delegate;
+
+namespace {
+
+Delegate * gDelegate = nullptr;
+
+Delegate * GetDelegate()
+{
+    if (gDelegate == nullptr)
+    {
+        static TimeSyncManager dg;
+        gDelegate = &dg;
+    }
+    return gDelegate;
+}
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace TimeSynchronization {
+
+void SetDefaultDelegate(Delegate * delegate)
+{
+    gDelegate = delegate;
+}
+
+Delegate * GetDefaultDelegate()
+{
+    return GetDelegate();
+}
+
+} // namespace TimeSynchronization
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+constexpr uint64_t kChipEpochUsSinceUnixEpoch = (uint64_t) kChipEpochSecondsSinceUnixEpoch * chip::kMicrosecondsPerSecond;
+
+static bool ChipEpochToUnixEpochMicro(uint64_t chipEpochTime, uint64_t & unixEpochTime)
+{
+    unixEpochTime = chipEpochTime + kChipEpochUsSinceUnixEpoch;
+    return true;
+}
+
+static bool UnixEpochToChipEpochMicro(uint64_t unixEpochTime, uint64_t & chipEpochTime)
+{
+    VerifyOrReturnValue(unixEpochTime >= kChipEpochUsSinceUnixEpoch, false);
+    chipEpochTime = unixEpochTime - kChipEpochUsSinceUnixEpoch;
+
+    return true;
+}
+
+static CHIP_ERROR UpdateUTCTime(uint64_t UTCTimeInChipEpochUs)
+{
+    System::Clock::Seconds32 lastKnownGoodChipEpoch;
+    uint64_t UTCTimeInUnixEpochUs;
+
+    ReturnErrorOnFailure(Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch));
+    VerifyOrReturnError(ChipEpochToUnixEpochMicro(UTCTimeInChipEpochUs, UTCTimeInUnixEpochUs), CHIP_ERROR_INVALID_TIME);
+
+    // if ( (UTCTimeInUnixEpochUs / chip::kMicrosecondsPerSecond) < lastKnownGoodChipEpoch) return CHIP_ERROR_INVALID_TIME;
```

Comment:
This is something I had initially which was preventing updating to an older Last Known Good Time. But the spec doesn't mandate it and it could be useful to change this if the wrong time is set. I will remove it.